### PR TITLE
[Minecraft] Prepare for Google Blockly Migration

### PR DESCRIPTION
**Pull request by mikeharv**
This branch offers a reasonably small number of changes that should effectively make Minecraft ready to test for migration to Google Blockly! 🎉 

There are quite a few files in this branch, but the changes are (hopefully) each fairly small and reasonable.

### Translations
The `repeat n times` block relies on translations that currently live in the CDO Blockly repo. I added new translations to our common locales. Since we might need to use either version of Blockly for now, I opted for setting the label like `.appendField(blocklyTranslation || commonTranslation)` since this would prioritize the Blockly repo's translations for CDO Blockly and give time for CrowdIn to catch up so we have fewer translation regressions on this block after flipping.

### Styles
In order for blocks to be compatible with our various Google Blockly themes, they need to have a set style property. CDO Blockly however uses a static HSV color value. Fortunately, Google Blockly prioritizes the style if it's present, so there's no harm in assigning both.
`Block.setStyle` doesn't normally exist on CDO Blockly, so I patched in the wrapper so that it doesn't cause any errors when called. I had previously added some pretty clunky logic where we were checking for the existence of this function to determine which Blockly version was in use. As a result of it now existing in both places, I had to clean up a couple of these places, such as in [`block_utils`](https://github.com/code-dot-org/code-dot-org/pull/55993/commits/7ee9d5ffdb738c0e2a52fa5378b74847c76e7112#diff-4204ee82808ca5d21acf4aef2f991c9674c29b599070f1c7aedc237c3ec3b461).
I needed to add styles to all the craft blocks (except for Aquatic, which were already completed year https://github.com/code-dot-org/code-dot-org/pull/50872), and some "common" blocks (e.g. repeat loops) shared across labs. Since I was already in the [`blocksCommon.js`](https://github.com/code-dot-org/code-dot-org/pull/55993/commits/7ee9d5ffdb738c0e2a52fa5378b74847c76e7112#diff-4ea1f53fc8a9f11e3683471951371663e4649e501aa020c9c06051be88db3ba8) file, I added styles for each block here, some of which are used more in Maze/Artist etc.

### Generators
For the v10 upgrade, we had to address a change that moves generator functions to a new `generator.forBlock` object. For our common blocks, I added some helper functions to the `customBlocks/*/commonBlocks.js` files that allow us to handle cases where we would set one block code generator to another, or assign a new function as a block code generator. 

_The changes described in the three sections above are mostly captured in the first commit: 7ee9d5ffdb738c0e2a52fa5378b74847c76e7112_

### Menu Generator for Repeat block Dropdown
I noticed a crash on any Minecraft level that used the dropdown variant of `repeat n times` block. This was due to not supplying the dropdown field with a menu generator when it was created. Instead these blocks set their options with a `config` attribute in the field element. For the Dance Party migration, we had to add support for this as some blocks used it to customize the menu options. https://github.com/code-dot-org/code-dot-org/pull/51750 However, these blocks also had a default menu generator which was supplied to the field. Since the blocks evidently work just fine with the field config, we can provide the dropdown field base class with a generic menu generator to avoid the error. 5a385056b3de4fa542e625702aa6d5609ca0fc39

### Uneditable Functions <a href="#uneditable-functions" id="uneditable-functions">#</a>
Some levels in the Agent (aka "hero") tutorial have function definitions on the primary workspace (no modal) with static call blocks defined in the toolbox XML. These definition blocks need to be uneditable, but they weren't. I believe this is similar to an issue the Music Lab team recently found with these blocks being movable. See thread: https://github.com/code-dot-org/code-dot-org/pull/55648#discussion_r1458134249
By adding a similar `initialEditConfig` property to the block's extra state, we can make sure that these blocks remain uneditable which prevents the toolbox blocks from breaking. 7057ec59b0cdce7ab053f0107eb8dfe5efdb9a6d

### Getting student code (Generators Part 2)
To actually run the student's code, we were calling `Generator.blocksToCode` with a list of block SVGs. This method is custom to CDO Blockly and unsupported. Fortunately, both versions of Blockly already have a working `getWorkspaceCode` function which effectively returns the same thing. You will notice that the old logic here also referenced `studioApp().initializationBlocks`. I checked every level at `dashboard/config/levels/custom/craft` and found that **none** them use initialization blocks, so I think it's safe to skip. 3c8902bd93dd86fc183976d7abf178ed19987f91

### "Invisible" Blocks
A few levels in the Minecraft Designer script include some invisible event blocks in the start blocks. Fortunately, we recently made the decision to continue supporting invisible top blocks by moving them to the hidden workspace: https://github.com/code-dot-org/code-dot-org/pull/55520
This requires all potentially-invisible blocks to have a generic set of serialization hooks that will save/load arbitrary properties (like `.invisible`). I added these hooks to the event blocks used throughout this tutorial. d0df44dccae7cc6dd8f3f6f2579c5d6e4f202270

### "Invisible" Blocks in a toolbox
A silly concept really. In exactly one level, we hide the `when run` in the workspace, but we also set the toolbox to contain two (seemingly random) invisible blocks. This is not something I think we need to support, and fortunately it seems to just impact one level. I update that level to remove the blocks instead. Since these toolbox blocks have never been visible, nobody will miss them. This does create a very small visual change, which is that Blockly no longer renders the flyout column to the right of the workspace. 6da10b0585172d86488769b5d361cc60c4a6c551
`/s/minecraft/lessons/1/levels/4` is shown below...
Before, with toolbox containing invisible blocks:
image
After, with no toolbox
image

## Testing story

I manually tested each level of each of the four Minecraft Hour of Code tutorials and spot-checked the cloned levels that appear in Courses D and E. I spot-checked levels in non-English, including checking that functions and validation worked as expected in the levels that require students to use blocks with translated procedure names.

I also manually checked for regressions in Minecraft using CDO Blockly. The biggest risk here was related to block colors and code generation, neither of which seems to be affected.

No unit tests are impacted. When/if we flip Minecraft (or Sprite Lab) we should anticipate some Eyes Tests will need to be updated.

## Deployment strategy

This branch doesn't actually flip the lab and shouldn't cause any regressions in Minecraft as long as it continues to run on CDO Blockly.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/i18n/common/en_us.json [hunk 1]
```diff
@@ -2138,7 +2138,7 @@
   "rubricLevelTwoHeader": "Convincing Evidence",
   "rubricLevelFourHeader": "No Evidence",
   "rubricScores": "Rubric Scores",
-  "rubricTabStudent" : "Student Rubric",
+  "rubricTabStudent": "Student Rubric",
```
Comment (mikeharv) [inline apps/i18n/common/en_us.json:2140]: Unrelated to PR. VSC automatically removed an extra space.

File: apps/src/block_utils.js [hunk 2]
```diff
@@ -1074,16 +1075,17 @@ exports.createJsWrapperBlockCreator = function (
...
-          if (this.setStyle) {
-            this.setStyle('default');
-          } else {
-            Blockly.cdoUtils.setHSV(this, ...DEFAULT_COLOR);
-          }
+          // CDO Blockly automatically sets block color for blocks with a specific
+          // returnType (e.g. yellow for "Location")
+          Blockly.cdoUtils.setHSV(this, ...DEFAULT_COLOR);
+        }
+        if (style) {
```
Comment (ebeastlake) [inline apps/src/block_utils.js]: nit: The syntax `this.setStyle(style ? style : 'default');` should also work.

File: apps/src/block_utils.js [hunk 3]
```diff
@@ -1074,16 +1075,17 @@ exports.createJsWrapperBlockCreator = function (
       helpUrl: getHelpUrl(docFunc), // optional param
       init: function () {
         // Styles should be used over hard-coded colors in Google Blockly blocks
-        if (style && this.setStyle) {
-          this.setStyle(style);
-        } else if (color) {
+        if (color) {
```
Comment (ebeastlake) [inline apps/src/block_utils.js:1084]: nit: I'd accept the argument that it's less clear and I'm not sure `chosenColor` is the right variable name, but could also do something like...
```
const chosenColor = color || (!returnType ? DEFAULT_COLOR : null);
if (chosenColor) {
    Blockly.cdoUtils.setHSV(this, ...chosenColor);
}
```

File: apps/src/blockly/cdoBlocklyWrapper.js [hunk 4]
```diff
@@ -257,6 +257,7 @@ function initializeBlocklyWrapper(blocklyInstance) {
   blocklyWrapper.Block.prototype.appendEndRowInput =
     blocklyWrapper.Block.prototype.appendDummyInput;
 
+  blocklyWrapper.Block.prototype.setStyle = function () {};
```
Comment (ebeastlake) [inline apps/src/blockly/cdoBlocklyWrapper.js:263]: Is it worth leaving a comment to the effect of the note in your PR description?

> Block.setStyle doesn't normally exist on CDO Blockly, so I patched in the wrapper so that it doesn't cause any errors when called.

File: apps/src/blockly/customBlocks/googleBlockly/mutators/procedureDefMutator.js [hunk 5]
```diff
@@ -160,6 +161,7 @@ export const procedureDefMutator = {
     setBlockDescription(this, state['description']);
     this.doProcedureUpdate();
     this.setDeletable(state['initialDeleteConfig'] === false ? false : true);
```
Comment (ebeastlake) [inline apps/src/blockly/customBlocks/googleBlockly/mutators/procedureDefMutator.js:163]: Does something like this make the intention of all the `state['initialEditConfig'] === false ? false : true` statements more explicit? 🤔  

```
function isExplicitlyFalse(value) {
    return value === false;
}
this.setDeletable(!isExplicitlyFalse(state['initialDeleteConfig']));
this.setEditable(!isExplicitlyFalse(state['initialEditConfig']));
this.setMovable(!isExplicitlyFalse(state['initialMoveConfig']));
this.setStatements_(!isExplicitlyFalse(state['hasStatements']));
```

File: apps/src/blocksCommon.js [hunk 6]
```diff
@@ -23,18 +30,27 @@ exports.install = function (blockly, blockInstallOptions) {
 
 function installControlsRepeatSimplified(blockly, skin) {
   // Re-uses the repeat block generator from core
-  blockly.JavaScript.controls_repeat_simplified =
-    blockly.JavaScript.controls_repeat;
-  blockly.JavaScript.controls_repeat_simplified_dropdown =
-    blockly.JavaScript.controls_repeat;
+  blockly.customBlocks.copyBlockGenerator(
```
Comment (ebeastlake) [inline apps/src/blocksCommon.js:26]: Did you add this util mainly so you don't have to write out
`blockly.JavaScript.forBlock.controls_repeat_simplified = blockly.JavaScript.forBlock.controls_repeat`?

File: apps/src/blocksCommon.js [hunk 7]
```diff
@@ -170,21 +208,22 @@ function installCategory(blockly) {
...
     // Block to handle event where mouse is clicked
     helpUrl: '',
     init: function () {
-      if (this.setStyle) {
-        this.setStyle('setup_blocks');
-      } else {
-        Blockly.cdoUtils.setHSV(this, 39, 1.0, 0.99);
-      }
+      Blockly.cdoUtils.setHSV(this, 39, 1.0, 0.99);
+      this.setStyle('setup_blocks');
```
Comment (ebeastlake) [inline apps/src/blocksCommon.js:219]: Is this logic the same because line 225 is what we want for CDO Blockly, in which case 226 is a no-op, and for Google Blockly, line 226 overrides line 255?

File: dashboard/config/levels/custom/craft/MC HOC 2016 New Walk.level [hunk 8]
```diff
@@ -68,12 +68,7 @@
       </xml>
     </start_blocks>
     <toolbox_blocks>
-      <xml>
```
Comment (ebeastlake) [inline dashboard/config/levels/custom/craft/MC HOC 2016 New Walk.level:71]: Was this removal intentional?

**Comment by ebeastlake** [comment]
I'm curious about answers to a couple of comments, but this looks good to me. Re-tagging @molly-moen because a second set of eyes never hurts 😊

File: apps/src/blocksCommon.js (hunk 6 above)
Comment (mikeharv) [reply apps/src/blocksCommon.js:26]: Only partially. If I had changed that directly this would crash in any CDO Blockly lab since forBlock is undefined.

File: apps/src/blocksCommon.js (hunk 7 above)
Comment (mikeharv) [reply apps/src/blocksCommon.js:219]: Yep!

**Review by molly-moen — APPROVED** [review]
I think this looks good, pending answers to Emily's comments/questions. For a future pr like this it would be good to split it up into a couple prs, I had a hard time context switching between the different changes (ex. one pr for the style changes, one for the generator changes, etc.).

Exciting that this was all that was needed to get Minecraft working!

File: apps/src/block_utils.js (hunk 2 above)
Comment (mikeharv) [reply apps/src/block_utils.js]: Or `this.setStyle(style || 'default');`?

Comment (ebeastlake) [reply apps/src/block_utils.js]: I get nervous about using the `||` operator like that for some reason, but I think it's unfounded. I think it's because I know the `?` operator forces the thing before it to be interpreted as either truthy or falsey, so it feels more readable to me in terms of explicitly showing the intent. I have more trouble remembering how JS treats falsy values with logical operators. But, I think you're right that they're the same here.

Screenshot 2024-01-26 at 5 11 29 PM

Now that I think about this a little more, I think I'm wary of `&&` and `||` because I'm overreacting to well-known bug in React where `0 && something` doesn't do what you want. Here's a link, just in case you're curious: https://www.codemzy.com/blog/react-render-0-conditional

File: apps/src/block_utils.js (hunk 3 above)
Comment (mikeharv) [reply apps/src/block_utils.js:1084]: I think we can't quite do that because setting the color to `null` (or any other negative value) would override the color that was previously set by the `returnType`. If there's a `returnType` at all, it's dangerous to call `.setHSV`. 

We could probably do something like this:
```
if (!returnType) {
  const blockColor = color || DEFAULT_COLOR;
  Blockly.cdoUtils.setHSV(this, ...blockColor);
}
```
but then that also slightly changes the behavior. This is probably an edge case to be fair. Right now a levelbuilder could specify a block to have both a `returnType` and a `color`, and the color would take precedence. That seems worth leaving in to me.
